File: vb/Platforms/Client.Wpf/MainViewModel.vb
Imports System
Imports Opc.UaFx
Imports Opc.UaFx.Client

Namespace Client.Wpf
    Public Class MainViewModel
        Inherits ViewModel

        '---------- Private fields ----------

        Private _address As Uri
        Private _addressStatus As String
        Private _client As OpcClient
        Private _subscription As OpcSubscription
        Private _nodeId As OpcNodeId
        Private _nodeStatus As String
        Private _nodeValue As Object

        '---------- Public constructors ----------

        Public Sub New()
            MyBase.New()

            Me._address = New Uri("opc.tcp://localhost:4840")
            Me._addressStatus = "Enter address of server."

            Me._nodeId = "ns=2;s=MyNode"
            Me._nodeStatus = "Enter a node identifier and click 'Subscribe'."

            Me.SubscribeCommand = New DelegateCommand(
                    execute:=Sub(__) Me.Subscribe(),
                    canExecute:=Function(__) Me.AddressIsValid AndAlso Me.NodeIdIsValid)
        End Sub

        '---------- Public properties ----------

        Public Property Address As String
            Get
                Return Me._address?.ToString()
            End Get
            Set(ByVal value As String)
                Me.RaisePropertyChanging(NameOf(Me.Address))
                Me.RaisePropertyChanging(NameOf(Me.AddressIsValid))

                Dim result As Uri = Nothing

                If Uri.TryCreate(value, UriKind.Absolute, result) _
                        AndAlso (result.Scheme = "opc.tcp" OrElse result.Scheme = "https") Then
                    Me._subscription?.Unsubscribe()
                    Me._client?.Disconnect()
                    Me._client = Nothing

                    Me._address = result

                    Me.RaisePropertyChanged(NameOf(Me.Address))
                    Me.RaisePropertyChanged(NameOf(Me.AddressIsValid))

                    Me.SubscribeCommand.RaiseCanExecuteChanged()
                Else
                    Me.AddressStatus = "Invalid address uri."
                End If
            End Set
        End Property

        Public ReadOnly Property AddressIsValid As Boolean
            Get
                Return Me.address IsNot Nothing
            End Get
        End Property

        Public Property AddressStatus As String
            Get
                Return Me._addressStatus
            End Get
            Set(ByVal value As String)
                Me.RaisePropertyChanging(NameOf(Me.AddressStatus))
                Me._addressStatus = value
                Me.RaisePropertyChanged(NameOf(Me.AddressStatus))
            End Set
        End Property

        Public Property NodeId As String
            Get
                Return Me._nodeId?.ToString()
            End Get
            Set(ByVal value As String)
                Me.RaisePropertyChanging(NameOf(Me.NodeId))
                Me.RaisePropertyChanging(NameOf(Me.NodeIdIsValid))

                Dim result = Nothing

                If OpcNodeId.TryParse(value, result) Then
                    Me._nodeId = result

                    Me.RaisePropertyChanged(NameOf(Me.NodeId))
                    Me.RaisePropertyChanged(NameOf(Me.NodeIdIsValid))

                    Me.SubscribeCommand.RaiseCanExecuteChanged()
                Else
                    Me.NodeStatus = "Invalid node identifier."
                End If
            End Set
        End Property

        Public ReadOnly Property NodeIdIsValid As Boolean
            Get
                Return Me._nodeId IsNot Nothing AndAlso Me._nodeId <> OpcNodeId.Null
            End Get
        End Property

        Public Property NodeStatus As String
            Get
                Return Me._nodeStatus
            End Get
            Set(ByVal value As String)
                Me.RaisePropertyChanging(NameOf(Me.NodeStatus))
                Me._nodeStatus = value
                Me.RaisePropertyChanged(NameOf(Me.NodeStatus))
            End Set
        End Property

        Public Property NodeValue As Object
            Get
                Return Me._nodeValue
            End Get
            Set(ByVal value As Object)
                Me.RaisePropertyChanging(NameOf(Me.NodeValue))
                Me._nodeValue = value
                Me.RaisePropertyChanged(NameOf(Me.NodeValue))
            End Set
        End Property

        Public ReadOnly Property SubscribeCommand As DelegateCommand

        '---------- Public methods ----------

        Public Sub Subscribe()
            Try
                If Me._client Is Nothing Then
                    Me._client = New OpcClient(Me.Address)
                    Me._client.Connect()
                End If

                If Me._subscription IsNot Nothing Then Me._subscription.Unsubscribe()

                Me._subscription = Me._client.SubscribeDataChange(
                        Me.NodeId,
                        AddressOf Me.HandleDataChangeReceived)

                Dim monitoredItem = Me._subscription.MonitoredItems(0)
                Me.NodeStatus = monitoredItem.Status.[Error].Description
            Catch ex As Exception
                Me.NodeStatus = ex.Message
            End Try
        End Sub

        '---------- Private methods ----------

        Private Sub HandleDataChangeReceived(ByVal sender As Object, ByVal e As OpcDataChangeReceivedEventArgs)
            Dim value = e.Item.Value

            Me.NodeStatus = value.Status.Description
            Me.NodeValue = $"{value.Value} ({e.MonitoredItem.NodeId})"
        End Sub
    End Class
End Namespace

File: cs/Platforms/Client.Wpf/MainViewModel.cs
namespace Client.Wpf
{
    using System;

    using Opc.UaFx;
    using Opc.UaFx.Client;

    public class MainViewModel : ViewModel
    {
        #region ---------- Private fields ----------

        private Uri address;
        private string addressStatus;

        private OpcClient client;
        private OpcSubscription subscription;

        private OpcNodeId nodeId;
        private string nodeStatus;
        private object nodeValue;

        #endregion

        #region ---------- Public constructors ----------

        public MainViewModel()
            : base()
        {
            this.address = new Uri("opc.tcp://localhost:4840");
            this.addressStatus = "Enter address of server.";

            this.nodeId = "ns=2;s=MyNode";
            this.nodeStatus = "Enter a node identifier and click 'Subscribe'.";

            this.SubscribeCommand = new DelegateCommand(
                    execute: (_) => this.Subscribe(),
                    canExecute: (_) => this.AddressIsValid && this.NodeIdIsValid);
        }

        #endregion

        #region ---------- Public properties ----------

        public string Address
        {
            get => this.address?.ToString();
            set
            {
                this.RaisePropertyChanging(nameof(this.Address));
                this.RaisePropertyChanging(nameof(this.AddressIsValid));

                if (Uri.TryCreate(value, UriKind.Absolute, out var result)
                        && (result.Scheme == "opc.tcp" || result.Scheme == "https")) {
                    this.subscription?.Unsubscribe();
                    this.client?.Disconnect();

                    this.client = null;
                    this.address = result;

                    this.RaisePropertyChanged(nameof(this.Address));
                    this.RaisePropertyChanged(nameof(this.AddressIsValid));

                    this.SubscribeCommand.RaiseCanExecuteChanged();
                }
                else {
                    this.AddressStatus = "Invalid address uri.";
                }
            }
        }

        public bool AddressIsValid
        {
            get => this.address != null;
        }

        public string AddressStatus
        {
            get => this.addressStatus;
            set
            {
                this.RaisePropertyChanging(nameof(this.AddressStatus));
                this.addressStatus = value;
                this.RaisePropertyChanged(nameof(this.AddressStatus));
            }
        }

        public string NodeId
        {
            get => this.nodeId?.ToString();
            set
            {
                this.RaisePropertyChanging(nameof(this.NodeId));
                this.RaisePropertyChanging(nameof(this.NodeIdIsValid));

                if (OpcNodeId.TryParse(value, out var result)) {
                    this.nodeId = result;

                    this.RaisePropertyChanged(nameof(this.NodeId));
                    this.RaisePropertyChanged(nameof(this.NodeIdIsValid));

                    this.SubscribeCommand.RaiseCanExecuteChanged();
                }
                else {
                    this.NodeStatus = "Invalid node identifier.";
                }
            }
        }

        public bool NodeIdIsValid
        {
            get => this.nodeId != null && this.nodeId != OpcNodeId.Null;
        }

        public string NodeStatus
        {
            get => this.nodeStatus;
            set
            {
                this.RaisePropertyChanging(nameof(this.NodeStatus));
                this.nodeStatus = value;
                this.RaisePropertyChanged(nameof(this.NodeStatus));
            }
        }

        public object NodeValue
        {
            get => this.nodeValue;
            set
            {
                this.RaisePropertyChanging(nameof(this.NodeValue));
                this.nodeValue = value;
                this.RaisePropertyChanged(nameof(this.NodeValue));
            }
        }

        public DelegateCommand SubscribeCommand
        {
            get;
        }

        #endregion

        #region ---------- Public methods ----------

        public void Subscribe()
        {
            try {
                if (this.client == null) {
                    this.client = new OpcClient(this.address);
                    this.client.Connect();
                }

                if (this.subscription != null)
                    this.subscription.Unsubscribe();

                this.subscription = this.client.SubscribeDataChange(
                        this.nodeId,
                        this.HandleDataChangeReceived);

                var monitoredItem = this.subscription.MonitoredItems[0];
                this.NodeStatus = monitoredItem.Status.Error.Description;
            }
            catch (Exception ex) {
                this.NodeStatus = ex.Message;
            }
        }

        #endregion

        #region ---------- Private methods ----------

        private void HandleDataChangeReceived(object sender, OpcDataChangeReceivedEventArgs e)
        {
            var value = e.Item.Value;

            this.NodeStatus = value.Status.Description;
            this.NodeValue = $"{value.Value} ({e.MonitoredItem.NodeId})";
        }

        #endregion
    }
}
